src/endpoint/s3/s3_utils.js

```diff
@@ -376,17 +376,17 @@ function parse_tagging_header(req) {
         dbg.error('parse_tagging_header failed', err);
         throw new S3Error(S3Error.MalformedXML);
     }
-    const tag_set = [];
-    const tag_map = new Map();
+    const tag_arr = [];
+    const tag_set = new Set();
     tagging_params.forEach((value, key) => {
         const tag = { key, value };
         if (!_is_valid_tag_values(tag)) throw new S3Error(S3Error.InvalidTag);
-        if (tag_map.has(tag.key)) throw new S3Error(S3Error.InvalidTag);
-        tag_map.set(tag.key);
-        tag_set.push(tag);
+        if (tag_set.has(tag.key)) throw new S3Error(S3Error.InvalidTag);
+        tag_set.add(tag.key);
+        tag_arr.push(tag);
```

Comment:
Instead of collecting both in a set and into an array, you can just add to the set and then `return Array.from(tag_set)`